### PR TITLE
RavenDB-8964

### PR DESCRIPTION
**Pull request by aviv86**
- refactoring
 - add support for First() and FirstOrDefault()
 - support nested queries with load
 - relevant tests

### CODE REVIEW

File: test/FastTests/Client/QueriesWithCustomFunctions.cs [hunk 1]
```diff
@@ -1668,6 +1668,112 @@ select new
...
+                                let details = RavenQuery.Load<Detail>(u.DetailIds)
+                                select new 
+                                {
+                                    Name = u.Name,
+                                    First = details.First(x=>x.Number > 1).Number,
+                                    FirstOrDefault = details.FirstOrDefault(x => x.Number < 3)
+                                };
+
+                    Assert.Equal("FROM Users as u LOAD u.DetailIds as details[] " +
+                                 "SELECT { Name : u.Name, First : details.filter(function(x){return x.Number>1;})[0].Number, " +
```
Comment (lahma) [inline test/FastTests/Client/QueriesWithCustomFunctions.cs]: Filter iterates whole collection and this gives array bounds error if not found? Find would have correct semantics if supported https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find .

File: src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs [hunk 2]
```diff
@@ -1569,15 +1564,35 @@ private void VisitSelect(Expression operand)
...
                         }
 
                         var expression = _linqPathProvider.GetMemberExpression(newExpression.Arguments[index]);
-                        AddToFieldsToFetch(GetSelectPath(expression), GetSelectPath(newExpression.Members[index]));
+
+                        try
+                        {
+                            AddToFieldsToFetch(GetSelectPath(expression), GetSelectPath(newExpression.Members[index]));
+                        }
+                        catch (Exception)
```
Comment (ayende) [inline src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs]: This is generally something that we don't want to do in this manner. Especially not catch a generic exception.
Handle the expect error here upfront.

File: src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs [hunk 3]
```diff
@@ -1609,17 +1624,36 @@ private void VisitSelect(Expression operand)
...
                         }
 
                         var expression = _linqPathProvider.GetMemberExpression(field.Expression);
-                        var renamedField = GetSelectPath(expression);
+                        try
+                        {
+                            var renamedField = GetSelectPath(expression);
+                            AddToFieldsToFetch(renamedField, GetSelectPath(field.Member));
+                        }
+                        catch (Exception)
```
Comment (ayende) [inline src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs]: Same

File: src/Raven.Client/Util/JavascriptConversionExtensions.cs [hunk 4]
```diff
@@ -125,6 +126,22 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
...
+                    case "FirstOrDefault":
+                    case "First":
+                        context.PreventDefault();
+                        context.Visitor.Visit(methodCallExpression.Arguments[0]);
+                        var writer = context.GetWriter();
+                        using (writer.Operation(methodCallExpression))
+                        {
+                            if (methodCallExpression.Arguments.Count > 1)
+                            {
+                                writer.Write(".filter");
```
Comment (ayende) [inline src/Raven.Client/Util/JavascriptConversionExtensions.cs]: Use `find` instead of `filter`

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Util/JavascriptConversionExtensions.cs (hunk 4 above)
Comment (aviv86) [reply src/Raven.Client/Util/JavascriptConversionExtensions.cs]: I tried using find, it seems like it's not supported in jint, that's why i went with filter..

running this query in studio :
from Regions as r
select {  T : r.Territories.find(function(x){return x !== null;}) }

gives this error :
Raven.Client.Exceptions.Documents.Patching.JavaScriptException: At 4:2
{"name":"TypeError","message":"Object has no method 'find'","callstack":" at find(Esprima.Ast.FunctionExpression) @  8:3\r\n at find(Esprima.Ast.FunctionExpression) @  8:3\r\n"} ---> Jint.Runtime.JavaScriptException: Object has no method 'find'
   at Jint.Native.Function.ScriptFunctionInstance.Call(JsValue thisArg, JsValue[] arguments)
   at Raven.Server.Documents.Patch.ScriptRunner.SingleRun.Run(DocumentsOperationContext ctx, String method, Object[] args) in C:\Builds\RavenDB-4.0-Nightly\src\Raven.Server\Documents\Patch\ScriptRunner.cs:line 426

Comment (ayende) [reply src/Raven.Client/Util/JavascriptConversionExtensions.cs]: Add this as polyfill, thne.

**Pull request merged**